src/transaction/log_replication.cpp

```diff
@@ -0,0 +1,171 @@
+    auto task_func = std::bind (&replicator::redo_upto_nxiolsa, std::ref (*this), std::placeholders::_1);
+    auto task = new cubthread::entry_callable_task (task_func);
+
+    m_daemon = cubthread::get_manager ()->create_daemon (loop, task, "cublog::replicator");
+  }
+
+  replicator::~replicator ()
+  {
+    cubthread::get_manager ()->destroy_daemon (m_daemon);
+
+    log_zip_free_data (m_undo_unzip);
+    log_zip_free_data (m_redo_unzip);
+  }
+
+  void
+  replicator::redo_upto_nxiolsa (cubthread::entry &thread_entry)
+  {
+    thread_entry.tran_index = LOG_SYSTEM_TRAN_INDEX;
+
+    while (true)
+      {
+	log_lsa nxio_lsa = log_Gl.append.get_nxio_lsa ();
+	if (m_redo_lsa < nxio_lsa)
+	  {
+	    redo_upto (thread_entry, nxio_lsa);
+	  }
+	else
+	  {
+	    assert (m_redo_lsa == nxio_lsa);
+	    break;
+	  }
+      }
+  }
+
+  void
+  replicator::redo_upto (cubthread::entry &thread_entry, const log_lsa &end_redo_lsa)
+  {
+    assert (m_redo_lsa < end_redo_lsa);
+
+    // redo all records from current position (m_redo_lsa) until end_redo_lsa
+
+    // make sure the log page is refreshed. otherwise it may be outdated and new records may be missed
+    m_reader.set_lsa_and_fetch_page (m_redo_lsa, log_reader::fetch_mode::FORCE);
+
+    while (m_redo_lsa < end_redo_lsa)
+      {
+	// read and redo a record
+	m_reader.set_lsa_and_fetch_page (m_redo_lsa);
+
+	log_rec_header header = m_reader.reinterpret_copy_and_add_align<log_rec_header> ();
+
+	switch (header.type)
+	  {
+	  case LOG_REDO_DATA:
+	    read_and_redo_record<log_rec_redo> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_MVCC_REDO_DATA:
+	    read_and_redo_record<log_rec_mvcc_redo> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_UNDOREDO_DATA:
+	  case LOG_DIFF_UNDOREDO_DATA:
+	    read_and_redo_record<log_rec_undoredo> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_MVCC_UNDOREDO_DATA:
+	  case LOG_MVCC_DIFF_UNDOREDO_DATA:
+	    read_and_redo_record<log_rec_mvcc_undoredo> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_RUN_POSTPONE:
+	    read_and_redo_record<log_rec_run_postpone> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_COMPENSATE:
+	    read_and_redo_record<log_rec_compensate> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_DBEXTERN_REDO_DATA:
+	  {
+	    log_rec_dbout_redo dbout_redo = m_reader.reinterpret_copy_and_add_align<log_rec_dbout_redo> ();
+	    log_rcv rcv;
+	    rcv.length = dbout_redo.length;
+	    log_rv_redo_record (&thread_entry, m_reader, RV_fun[dbout_redo.rcvindex].redofun, &rcv, &m_redo_lsa, 0,
+				nullptr, m_redo_unzip);
+	  }
+	  break;
+	  default:
+	    // do nothing
+	    break;
+	  }
+
+	m_redo_mutex.lock ();
+	m_redo_lsa = header.forw_lsa;
+	m_redo_mutex.unlock ();
+	if (m_redo_lsa == end_redo_lsa)
+	  {
+	    // notify who waits for end of replication
+	    m_redo_condvar.notify_all ();
```

Comment:
I think the appropriate context to use notify_one() is when the notifier means that only one waiter must be awoken even if there are multiple waiters. An example is a producer/consumer queue. Whenever a new item is produced, a single consumer has to be notified even if there are more waiters.

I don't think notify_one() should be used because we are sure there is only one waiter. First reason is because we cannot make an assertion that there is indeed only one waiter. Second reason is that in this context, it matters not to the notifier if there is only one waiter or multiple waiters, it should notify them all.

If we add another wait_for_replication call in the future, it may be easily missed that the second waiter is not going to be notified.

Reply:
Agreed